### PR TITLE
Token Paymasters

### PR DESCRIPTION
**Pull request by smcsicardi**

### CODE REVIEW

File: contracts/paymaster/DeployPaymaster.sol [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    virtual
+    relayHubOnly
+    returns (bytes memory context, bool revertOnRecipientRevert) {
+        require(tokens[relayRequest.request.tokenContract], "Token contract not allowed");
+        IERC20 token = IERC20(relayRequest.request.tokenContract);
+
+        require(address(relayRequest.request.factory) != address(0), "factory should be defined");
+        require(address(relayRequest.request.factory) == factory, "factory should be the trusted one!");
+        
+        address contractAddr = ProxyFactory(relayRequest.request.factory).getSmartWalletAddress(relayRequest.request.from, relayRequest.request.to, relayRequest.request.data);
```
Comment (raullaprida) [inline contracts/paymaster/DeployPaymaster.sol:47]: a new PR to master added new parameters, "Index" and "recoverer" to getSmartWalletAddress

Comment (raullaprida) [reply contracts/paymaster/DeployPaymaster.sol:47]: As soon as it gets approved, your PR wont merge

Comment (raullaprida) [reply contracts/paymaster/DeployPaymaster.sol:47]: https://github.com/rsksmart/enveloping/pull/33

File: contracts/paymaster/RelayPaymaster.sol [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    relayHubOnly
+    returns (bytes memory context, bool revertOnRecipientRevert) {
+        address payer = relayRequest.relayData.forwarder;
+        require(tokens[relayRequest.request.tokenContract], "Token contract not allowed");
+        IERC20 token = IERC20(relayRequest.request.tokenContract);
+        uint256 tokenAmount = relayRequest.request.tokenAmount;
+        require(tokenAmount <= token.balanceOf(payer), "balance too low");
+
+        // Check for the codehash of the smart wallet sent
+        {
```
Comment (raullaprida) [inline contracts/paymaster/RelayPaymaster.sol:52]: what is the purpose of this new scope?

File: contracts/factory/ProxyFactory.sol [hunk 3]
```diff
@@ -94,6 +102,8 @@ contract ProxyFactory is IProxyFactory {
      */
     constructor(address forwarderTemplate) public {
         masterCopy = forwarderTemplate;
+
+        runtimeCodeHash = keccak256(abi.encodePacked(RUNTIME_START, masterCopy, RUNTIME_END));
```
Comment (raullaprida) [inline contracts/factory/ProxyFactory.sol:106]: keccak256(abi.encodePacked(RUNTIME_START, forwarderTemplate, RUNTIME_END))

File: contracts/factory/ProxyFactory.sol [hunk 4]
```diff
@@ -86,6 +88,12 @@ contract ProxyFactory is IProxyFactory {
     string public constant FORWARDER_PARAMS = "address from,address to,uint256 value,uint256 gas,uint256 nonce,bytes data,address tokenRecipient,address tokenContract,uint256 tokenAmount,address factory";
     string public constant EIP712_DOMAIN_TYPE = "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)";
 
+    bytes private constant CONSTR = hex"602D3D8160093D39F3";
```
Comment (raullaprida) [inline contracts/factory/ProxyFactory.sol:91]: bytes9

File: contracts/factory/ProxyFactory.sol [hunk 5]
```diff
@@ -86,6 +88,12 @@ contract ProxyFactory is IProxyFactory {
...
+    bytes private constant RUNTIME_START = hex"363D3D373D3D3D3D363D73";
```
Comment (raullaprida) [inline contracts/factory/ProxyFactory.sol:92]: bytes11

File: contracts/factory/ProxyFactory.sol [hunk 6]
```diff
@@ -86,6 +88,12 @@ contract ProxyFactory is IProxyFactory {
...
+    bytes private constant RUNTIME_END = hex"5AF43D923D90803E602B57FD5BF3";
```
Comment (raullaprida) [inline contracts/factory/ProxyFactory.sol:93]: bytes14

**Review by raullaprida — CHANGES_REQUESTED** [review]
Good progress!.
See my comments regarding PR 33 which will generate some conflicts with your PR (PR33 is already reviewed it will most likely be merged before this one)

File: contracts/paymaster/DeployPaymaster.sol (hunk 1 above)
Comment (smcsicardi) [reply contracts/paymaster/DeployPaymaster.sol]: fixed

**Review by raullaprida — APPROVED** [review]
LGTM!

**Pull request merged**